### DISCUSSION TITLE
Stripe Webhook return res.status(200) AND run logic

### DISCUSSION THREAD

**Discussion by fstrauf** [vercel/next.js#49282] [Help]
### Summary

Hi pros!

I'm trying to connect the stripe webhook to my serverless function.

I need to run a bit of logic to save the stripe-webhook-info to my db. It won't be fast enough if I send the 200 after I save to my db. Here is what I get: 

`An error occurred with your deployment
FUNCTION_INVOCATION_TIMEOUT`

I tried making my db call async (without waiting for its return), but it seems that once I `res.status(200).send()` it kills all async calls. 

All this being said, there must be someone out there that has solved this problem.

My current webhook is pretty much stock standard as per stripes docs.

How did you do this? Send a quick 200 to Stripe AND process a small db transaction.

### Additional information

```js
import type { NextApiRequest, NextApiResponse } from 'next'
import Stripe from 'stripe'
import { buffer } from 'micro'
import prisma from '../../lib/prisma'

export const config = {
  api: {
    bodyParser: false,
  },
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method === 'POST') {
    const stripe = new Stripe(process.env.STRIPE_SECRET_KEY, {
      apiVersion: '2022-11-15',
    })
    const sig = req.headers['stripe-signature']
    const buf = await buffer(req)
    let event
    try {
      event = await stripe.webhooks.constructEvent(
        buf,
        sig,
        process.env.STRIPE_WEBHOOK_SECRET
      )
      // res.status(200)
      switch (event?.type) {
        case 'checkout.session.completed':
          const userId = event.data.object?.client_reference_id
         
          const checkoutSessionId = event.data.object?.id
          const customer = event.data.object?.customer
          let productTier = ''
          try {
            const checkoutSession = await stripe.checkout.sessions.retrieve(
              checkoutSessionId,
              {
                expand: ['line_items'],
              }
            )
            productTier = String(
              checkoutSession.line_items.data[0].price.product
            )
          } catch (error) {
            console.log('🚀 ~ file: stripeHook.ts:53 ~ error:', error)
            //do nothing
          }
          
          try {
            prisma.subscriptions.upsert({
              where: {
                authorClerkId: userId,
              },
              create: {
                authorClerkId: userId,
                stripeCustomerId: customer,
                tier: productTier,
              },
              update: {
                stripeCustomerId: customer,
                tier: productTier,
              },
            })
          } catch (error) {
            console.error(error)
            console.log('prisma before')
            res.status(400).json({ error: `Webhook Error: ${err.message}` })
          }
          res.status(200).send({event: event?.type})
          //update the user publicmetadata with the new subscription data.
          break
        default:
          console.log(`Unhandled event type ${event?.type}`)
          res.status(200).send({event: event?.type})
      }

      res.status(200)
    } catch (err) {
      console.error(`Error verifying Stripe webhook: ${err.message}`)
      res.status(400).json({ error: `Webhook Error: ${err.message}` })
    }
  } else {
    console.log('only post allowed')
    res.setHeader('Allow', 'POST')
    res.status(405).end('Method Not Allowed')
  }
}
```

**Comment by Ilia01** [comment]
The main constraint here is that once you send the `200`, you should assume the serverless invocation can be frozen or terminated at any point. So "return 200 and keep doing work in the same request" is not something I would rely on.

A safer pattern is:

1. Verify the Stripe signature.
2. Persist the minimum durable state, or enqueue a job, keyed by `event.id`.
3. Return `200`.
4. Do the heavier side effects in a separate worker/background process.

A couple of details from your snippet:

- `prisma.subscriptions.upsert(...)` is missing `await`, so that `try/catch` will not see failures.
- webhook handlers should be idempotent; use `event.id` or your own uniqueness constraint so retries do not double-apply the mutation.
- if the DB write is actually tiny, it is often simpler to `await` that single write and return `200` only after it succeeds.

So I would not optimize for "send 200 immediately no matter what". I would optimize for "ack only after I have durably recorded the event or queued the work".

Separate point: if you are debugging raw-body or signature failures locally before you even get to this stage, I built HookLens for that exact inspect/replay loop.
